OpenVSX proxy cache

corneliusludmann opened the pull request:
This PR adds an OpenVSX proxy cache that caches OpenVSX requests for 30 seconds and holds old OpenVSX caches for 3 days in the cache and uses old caches in case the upstream OpenVSX answers with a 5xx HTTP status. 

This configuration holds the caches for 3 days. There is a risk when we bring this into production that the cache does not scale when holding the responses of the last 3 days (e.g. cache file size).

This change introduces the following new `values.yaml` configuration values:
```
vsxRegistry:
  proxy: false
  host: open-vsx.org
```
The proxy cache is activated for preview environments.

## How to test

Testing is pretty complex since one cannot simply stop open-vsx.org. I tested the changes with the following steps:
1. Start a reverse proxy somewhere that routes the traffic to open-vsx.org.
<details>
  <summary>click here for a python example</summary>
  
```python
#!/usr/bin/env python3

from http.server import BaseHTTPRequestHandler,HTTPServer
import argparse, os, random, sys, requests

from socketserver import ThreadingMixIn
import threading

hostname = 'en.wikipedia.org'

def merge_two_dicts(x, y):
    z = x.copy()   # start with x's keys and values
    z.update(y)    # modifies z with y's keys and values & returns None
    return z

def set_header():
    headers = {
        'Host': hostname
    }

    return headers

class ProxyHTTPRequestHandler(BaseHTTPRequestHandler):
    protocol_version = 'HTTP/1.0'
    def do_HEAD(self):
        self.do_GET(body=False)
        return
        
    def do_GET(self, body=True):
        sent = False
        try:

            url = 'https://{}{}'.format(hostname, self.path)
            req_header = self.parse_headers()

            print(req_header)
            print(url)
            resp = requests.get(url, headers=merge_two_dicts(req_header, set_header()), verify=False)
            sent = True

            self.send_response(resp.status_code)
            self.send_resp_headers(resp)
            msg = resp.text
            # msg = msg.replace('Welcome','CHANGED HERE') 
            if body:
                self.wfile.write(msg.encode(encoding='UTF-8',errors='strict'))
            return
        finally:
            if not sent:
                self.send_error(404, 'error trying to proxy')

    def do_POST(self, body=True):
        sent = False
        try:
            url = 'https://{}{}'.format(hostname, self.path)
            content_len = int(self.headers.get('content-length', 0))
            post_body = self.rfile.read(content_len)
            req_header = self.parse_headers()

            resp = requests.post(url, data=post_body, headers=merge_two_dicts(req_header, set_header()), verify=False)
            sent = True

            self.send_response(resp.status_code)
            self.send_resp_headers(resp)
            if body:
                self.wfile.write(resp.content)
            return
        finally:
            if not sent:
                self.send_error(404, 'error trying to proxy')

    def parse_headers(self):
        req_header = {}
        print(self.headers)
        for line in self.headers:
            req_header[line] = self.headers.get(line)
        print(req_header)
        return req_header

    def send_resp_headers(self, resp):
        respheaders = resp.headers
        print ('Response Header')
        for key in respheaders:
            if key not in ['Content-Encoding', 'Transfer-Encoding', 'content-encoding', 'transfer-encoding', 'content-length', 'Content-Length']:
                print (key, respheaders[key])
                self.send_header(key, respheaders[key])
        self.send_header('Content-Length', len(resp.content))
        self.end_headers()

def parse_args(argv=sys.argv[1:]):
    parser = argparse.ArgumentParser(description='Proxy HTTP requests')
    parser.add_argument('--port', dest='port', type=int, default=9999,
                        help='serve HTTP requests on specified port (default: random)')
    parser.add_argument('--hostname', dest='hostname', type=str, default='en.wikipedia.org',
                        help='hostname to be processd (default: en.wikipedia.org)')
    args = parser.parse_args(argv)
    return args

class ThreadedHTTPServer(ThreadingMixIn, HTTPServer):
    """Handle requests in a separate thread."""

def main(argv=sys.argv[1:]):
    global hostname
    args = parse_args(argv)
    hostname = args.hostname
    print('http server is starting on {} port {}...'.format(args.hostname, args.port))
    server_address = ('', args.port)
    httpd = ThreadedHTTPServer(server_address, ProxyHTTPRequestHandler)
    print('http server is running as reverse proxy')
    httpd.serve_forever()

if __name__ == '__main__':
    main()
```

- run with: `python proxy.py --hostname open-vsx.org --port 8080`
- based on: https://github.com/humbertodias/python-reverse-proxy
</details>

2. Edit the proxy-config configmap and change `to https://open-vsx.org` to the address of your reverse proxy. Restart proxy pod.
3. Start a workspace. In the proxy logs you'll find entries with `cacheStatus: miss`, plugins are installed fine
4. Stop the reverse proxy and start a webserver that answers with a status code >= 500 instead on the same host.
<details>
  <summary>click here for a python example</summary>
  
```python
#!/usr/bin/env python3

from http.server import BaseHTTPRequestHandler, HTTPServer
import time

hostName = ""
serverPort = 8080

class MyServer(BaseHTTPRequestHandler):
    def do_GET(self):
        self.send_response(501)
        self.send_header("Content-type", "text/html")
        self.end_headers()
        self.wfile.write(bytes("<html><head><title>https://pythonbasics.org</title></head>", "utf-8"))
        self.wfile.write(bytes("<p>Request: %s</p>" % self.path, "utf-8"))
        self.wfile.write(bytes("<body>", "utf-8"))
        self.wfile.write(bytes("<p>This is an example web server.</p>", "utf-8"))
        self.wfile.write(bytes("</body></html>", "utf-8"))

    def do_POST(self):
        content_length = int(self.headers['Content-Length']) # <--- Gets the size of data
        post_data = self.rfile.read(content_length) # <--- Gets the data itself
        print(post_data)
        self.send_response(501)
        self.send_header("Content-type", "text/html")
        self.end_headers()
        self.wfile.write(bytes("<html><head><title>https://pythonbasics.org</title></head>", "utf-8"))
        self.wfile.write(bytes("<p>Request: %s</p>" % self.path, "utf-8"))
        self.wfile.write(bytes("<body>", "utf-8"))
        self.wfile.write(bytes("<p>This is an example web server.</p>", "utf-8"))
        self.wfile.write(bytes("</body></html>", "utf-8"))


if __name__ == "__main__":        
    webServer = HTTPServer((hostName, serverPort), MyServer)
    print("Server started http://%s:%s" % (hostName, serverPort))

    try:
        webServer.serve_forever()
    except KeyboardInterrupt:
        pass

    webServer.server_close()
    print("Server stopped.")
```

- run with: `python proxy.py`
</details>

5. Start another workspace for the same repo. All plugins should be installed and the proxy logs `cacheStatus: hit`.


## Limitations

This cache works only for 5xx responses. If the server does not answer at all, it seems not to work. However, in the past when OpenVSX was down it usually sent 5xx responses from the load balancer.

## Next steps
- Add the config to the docs.
- Activate the proxy cache for staging / production.

---

Fixes #4904 and probably fixes #5031

- [x] /werft with-clean-slate-deployment

roboquat commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/4997#" title="Author self-approved">corneliusludmann</a>*

*No associated issue*. Update pull-request body to add a reference to an issue, or get approval with `/approve no-issue`

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)~~ [corneliusludmann]
- ~~[components/proxy/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/proxy/OWNERS)~~ [corneliusludmann]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

corneliusludmann commented:
If someone would like to take this over while I'm on vacation, please go ahead. Otherwise, I will continue this after my vacation.

/cc @csweichel @akosyakov

roboquat commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **corneliusludmann** after the PR has been reviewed.

Associated issue: *#4904*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)**
- **[components/proxy/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/proxy/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["corneliusludmann"]} -->

roboquat commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **csweichel** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @csweichel` in a comment when ready.

Associated issue: *#4904*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)**
- **[components/ide/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ide/OWNERS)**
- **[components/proxy/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/proxy/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["csweichel"]} -->

akosyakov commented on components/ide/code/startup.sh:
```diff
@@ -33,6 +33,10 @@ export USER=gitpod
 cd /ide || exit
+
+# Replace OpenVSX URL
+sed -i "s|https://open-vsx.org|$VSX_REGISTRY_URL|g" ./product.json
```
Did you switch to latest VS Code in settings? 😬

corneliusludmann replied:
> Did you switch to latest VS Code in settings?

Exactly that was the missing piece! :roll_eyes:  Thanks, @akosyakov for the hint!

corneliusludmann commented on components/proxy/Dockerfile:
```diff
@@ -25,7 +25,8 @@ RUN xcaddy build \
   --with github.com/gitpod-io/gitpod/proxy/plugins/logif=/plugins/logif \
-  --with github.com/gitpod-io/gitpod/proxy/plugins/jsonselect=/plugins/jsonselect
+  --with github.com/gitpod-io/gitpod/proxy/plugins/jsonselect=/plugins/jsonselect \
+  --with github.com/sillygod/cdp-cache
```
Pin specific version?

corneliusludmann commented on components/proxy/plugins/jsonselect/plugin.go:
```diff
@@ -104,6 +104,10 @@ func (e JSONSelectEncoder) EncodeEntry(entry zapcore.Entry, fields []zapcore.Fie
 	// TODO(leo): switch back to EachKey (see git history) for perf reasons when fixed
 	for idx, paths := range e.getters {
 		val, typ, _, err := jsonparser.Get(buf.Bytes(), paths...)
+		if err == jsonparser.KeyPathNotFoundError {
```
We need this check to handle access to headers that are not available in all requests, e.g.:

```
{cacheStatus:resp_headers>X-Cache-Status>[0]}
```

Without this check, this would lead to an error when the header `X-Cache-Status` is missing.

/cc @leodido

corneliusludmann commented on chart/templates/proxy-configmap.yaml:
```diff
@@ -51,6 +51,49 @@ data:
+
+      http_cache {
+        cache_type file
+        path /tmp/openvsx-cache
```
Is this a good place for the cache files?

akosyakov commented on chart/templates/blobserve-configmap.yaml:
```diff
@@ -32,6 +32,11 @@ data:
                     "replacements": [
                         { "search": "vscode-webview.net", "replacement": "{{ .Values.hostname }}", "path": "/ide/out/vs/workbench/workbench.web.api.js" },
```
this line does not seem to be appied anymore

<img width="482" alt="Screenshot 2021-08-31 at 15 48 01" src="https://user-images.githubusercontent.com/3082655/131514186-3314dbdd-1e31-40be-ab76-85d182dd4ebb.png">

You can reproduce it with `Simple Browser: Show` command.

akosyakov replied:
@corneliusludmann I tried in another prev env and could not reproduce it:
<img width="1904" alt="Screenshot 2021-09-03 at 08 37 18" src="https://user-images.githubusercontent.com/3082655/131961352-8c965354-d2b4-440d-885f-79dd61124eaa.png">

I think it happens only in this PR and caused by some bug in the blobserve replacement logic.

corneliusludmann replied:
Indeed, the paths are stored as key in a map so that the second entry replaces the first one.

https://github.com/gitpod-io/gitpod/blob/5c73a864d94cc9436cdb9341b6f1803ea2cd7a81/components/blobserve/pkg/blobserve/refstore.go#L29-L31

corneliusludmann replied:
@akosyakov Fixed in 69a8c54e727da6f1fcf483d63e38bff77029527b.

corneliusludmann commented:
@akosyakov figured out that there are OpenVSX URLs in the response bodies that need to be replaced as well. With commit 96b3a22 a Caddy plugin does this.

roboquat commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **csweichel** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @csweichel` in a comment when ready.

Associated issue: *#4904*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)**
- **[components/blobserve/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/blobserve/OWNERS)**
- **[components/ide/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ide/OWNERS)**
- **[components/proxy/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/proxy/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["csweichel"]} -->

akosyakov commented:
/werft run

akosyakov commented:
I tried to search for python and vim extensions and install them, but first failed to be found with `We cannot connect to the extension marketplace` and second failed to be installed.

corneliusludmann commented:
> I tried to search for python and vim extensions and install them, but first failed to be found with `We cannot connect to the extension marketplace` and second failed to be installed.

Thanks, @akosyakov for finding these 2 issues. This concerns two cases that did not occur in my tests:
1. When the result set is pretty large the body interception didn't work (that's for the `python` case; in my tests, the search term was always more narrow with fewer results). I fixed this.
2. There is a CORS issue with the preflight OPTIONS call that we serve statically in the proxy. That happens only for some extensions, e.g. the `vim` extension they call something like `https://open-vsx.org/vscode/asset/vscodevim/vim/1.21.6/Microsoft.VisualStudio.Code.WebResources/extension/package.json`. I added the following headers to the response and this fixed it:
```
Access-Control-Allow-Credentials "true"
Access-Control-Allow-Headers "content-type,x-market-client-id,x-market-user-id,x-client-commit,x-client-name,x-client-version,x-machine-id"
Access-Control-Allow-Methods "OPTIONS,GET,POST,PATCH,PUT,DELETE"
Access-Control-Allow-Origin "*"
```

Please give it another try!

akosyakov commented:
@corneliusludmann everything seems to work now

/approve
/assign @csweichel

roboquat commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/4997#issuecomment-915137565" title="Approved">akosyakov</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **csweichel** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @csweichel` in a comment when ready.

Associated issue: *#4904*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)~~ [akosyakov]
- **[components/blobserve/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/blobserve/OWNERS)**
- ~~[components/ide/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ide/OWNERS)~~ [akosyakov]
- **[components/proxy/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/proxy/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["csweichel"]} -->

csweichel commented:
/approve

roboquat commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/4997#issuecomment-915137565" title="Approved">akosyakov</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/4997#issuecomment-915869391" title="Approved">csweichel</a>*

Associated issue: *#4904*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)~~ [akosyakov,csweichel]
- ~~[components/blobserve/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/blobserve/OWNERS)~~ [csweichel]
- ~~[components/ide/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ide/OWNERS)~~ [akosyakov,csweichel]
- ~~[components/proxy/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/proxy/OWNERS)~~ [csweichel]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [csweichel]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

JanKoehnlein commented:
/lgtm

roboquat commented:
LGTM label has been added.  <details>Git tree hash: 42ecb975de397436601c23906aa55d4a71a08a58</details>

roboquat commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/4997#issuecomment-915137565" title="Approved">akosyakov</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/4997#issuecomment-915869391" title="Approved">csweichel</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/4997#issuecomment-916713458" title="LGTM">JanKoehnlein</a>*

Associated issue: *#4904*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)~~ [JanKoehnlein,akosyakov,csweichel]
- ~~[components/blobserve/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/blobserve/OWNERS)~~ [csweichel]
- ~~[components/ide/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ide/OWNERS)~~ [akosyakov,csweichel]
- ~~[components/proxy/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/proxy/OWNERS)~~ [JanKoehnlein,csweichel]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [JanKoehnlein,csweichel]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->
